File: backend/app/models/lecture.py
import uuid
import enum
from datetime import datetime

from sqlalchemy import String, DateTime, func, ForeignKey, Enum
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class LectureStatus(str, enum.Enum):
    recording = "recording"
    processing = "processing"
    completed = "completed"


class Lecture(Base):
    __tablename__ = "lectures"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    owner_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"))
    title: Mapped[str] = mapped_column(String(255), default="Untitled Lecture")
    status: Mapped[LectureStatus] = mapped_column(
        Enum(LectureStatus), default=LectureStatus.recording
    )
    audio_url: Mapped[str | None] = mapped_column(String(1024), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    ended_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    owner: Mapped["User"] = relationship(back_populates="lectures")
    segments: Mapped[list["TranscriptSegment"]] = relationship(
        back_populates="lecture", cascade="all, delete-orphan", order_by="TranscriptSegment.start_ms"
    )
    notes: Mapped[list["Note"]] = relationship(back_populates="lecture", cascade="all, delete-orphan")
    quiz_questions: Mapped[list["QuizQuestion"]] = relationship(
        back_populates="lecture", cascade="all, delete-orphan"
    )
    chat_messages: Mapped[list["ChatMessage"]] = relationship(
        back_populates="lecture", cascade="all, delete-orphan"
    )

File: src/types/lecture.ts
export type LectureStatus = "recording" | "processing" | "completed";

export interface Lecture {
  id: string;
  title: string;
  status: LectureStatus;
  audio_url: string | null;
  created_at: string;
  ended_at: string | null;
}

export interface TranscriptSegment {
  id: string;
  text: string;
  start_ms: number;
  end_ms: number;
  is_final: boolean;
  created_at: string;
}

export interface LectureDetail extends Lecture {
  segments: TranscriptSegment[];
}

export interface Note {
  id: string;
  content: string;
  created_at: string;
}

// WebSocket event payloads pushed from the backend
export type LectureSocketEvent =
  | {
      type: "transcript";
      text: string;
      is_final: boolean;
      start_ms: number;
      end_ms: number;
    }
  | { type: "notes"; content: string; created_at: string }
  | { type: "error"; detail: string };
